docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,151 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability to extend the DNS configuration of the gateway with the DNSPolicy resource. This resource allows 
+users to configure health checks. As a result of configuring health checks, the controller creates the health checks in 
+Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it 
+becomes unhealthy, and enabling it again when it becomes healthy again.
+
+This feature has a few shortfalls:
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, 
+or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+3. requires using the gateway controller to implement, maintain and test multiple providers
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+
+### Nongoals
+* Ability for the health checks to reach endpoints in separate private networks
+* Transparently keep support for other health check providers like Route53
+* Having health checks for wildcard listeners
+
+## Use-cases
+* As a gateway administrator, I would like to define a health check that each service sitting behind a particular 
+listener across the production clusters has to implement to ensure we can automatically respond, failover and 
+mitigate a failing instance of the service
+
+## Proposal
+
+Currently, this functionality will be added to the existing MGC, and executed within that component. This will be created
+with the knowledge that it may need to be made into an external component in the future.
+
+#### `DNSPolicy` resource
+
+The presence of the `healthCheck` means that for every DNS endpoint (that is either an A record, or a CNAME to an external host), 
+a health check is created based on the health check configuration in the DNSPolicy.
+
+A `failureThreshold` field will be added to the health spec, allowing users to configure a number of consecutive health 
+check failures that must be observed before the endpoint is considered unhealthy.
+
+Example DNS Policy with a defined health check.
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+metadata:
+  name: prod-web
+  namespace: multi-cluster-gateways
+spec:
+  healthCheck:
+    endpoint: /health
+    failureThreshold: 5
+    port: 443
+    protocol: https
+    additionalHeaders:
+      - name: "..."
+        value: "..."
+      - name: "..."
+        value: "..."
```

Comment:
For the use case of setting an auth header in the request we should allow the value of the header to be retrieved from a secret, otherwise we make it harder for a user to keep their custom resources under version control.

Reply:
This is a great suggestion, thanks Roi. I've updated the proposal.